src/hardware/driver/dac.rs

```diff
@@ -0,0 +1,240 @@
+            ChannelVariant::LowNoiseCathodeGrounded => (Self::R_OUT_LN, false),
+            ChannelVariant::HighPowerAnodeGrounded => (Self::R_OUT_HP, true),
+            ChannelVariant::HighPowerCathodeGrounded => (Self::R_OUT_HP, false),
+        };
+        let mut dac_code = (r_out
+            * current
+            * (DacCode::MAX_DAC_WORD as f32 / DacCode::VREF_DAC))
+            as i32;
+
+        if is_inverted {
+            // Convert to inverted dac output for anode grounded Driver channel variants.
+            // These variants need (VREF_DAC - V_CURR) to produce the current CURR.
+            // No bitinversion and masking so we can still detect bounds errors.
+            dac_code = DacCode::MAX_DAC_WORD - dac_code - 1; // For now just subtract a bit to stay in bounds for 0 A
+        }
+
+        if !(0..DacCode::MAX_DAC_WORD).contains(&dac_code) {
+            return Err(Error::Bounds);
+        };
+
+        Ok(Self(dac_code as u32))
+    }
+}
+
+impl From<DacCode> for u32 {
+    fn from(code: DacCode) -> u32 {
+        code.0
+    }
+}
+
+pub struct Dac<SPI: Transfer<u8> + Write<u8>> {
+    spi: SPI,
+    sync_n: gpio::ErasedPin<gpio::Output>,
+    channel: ChannelVariant,
+}
+
+impl<SPI, E> Dac<SPI>
+where
+    SPI: Transfer<u8, Error = E> + Write<u8, Error = E>,
+    E: Debug,
+{
+    /// Construct a new [Dac] in the following state:
+    /// - reset
+    /// - 10 V plusminus 1.25 V refernece span configured
+    /// - temperature drift calibration performed
+    /// - deglitch circuit is enabled by default (fast-settling off)
+    /// - asyncronous output update (when sync goes high) by default
+    /// - 0.5-MHz max DAC update rate by default for enhanced THD performance
+    /// - TNH masked for code jump > 2^14 (default)
+    pub fn new(
+        spi: SPI,
+        sync_n: gpio::ErasedPin<gpio::Output>,
+        channel: ChannelVariant,
+    ) -> Self {
+        let mut dac = Dac {
+            spi,
+            sync_n,
+            channel,
+        };
+
+        // reset DAC
+        dac.write(TRIGGER::SRST::RESET, DAC_ADDR::TRIGGER);
+        // set to 10 V plusminus 1.25 V referenece span
+        dac.write(CONFIG1::VREFVAL::SPAN_10V_1_25V, DAC_ADDR::CONFIG1);
+        // perform calibration
+        dac.calibrate();
+
+        dac
+    }
+
+    /// Perform temperature drift calibration (waits until calibration is done).
+    pub fn calibrate(&mut self) {
+        self.write(CONFIG1::EN_TMP_CAL::ENABLED, DAC_ADDR::CONFIG1);
+        self.write(TRIGGER::RCLTMP::RECAL, DAC_ADDR::TRIGGER);
+        // continuously read recalibration done bit until it is set
+        while (self.read(DAC_ADDR::STATUS) & STATUS::ALM::RECALIBRATED) == 0 {}
+    }
+
+    /// Set the DAC to produce a voltage corresponding to `current`.
+    pub fn set(&mut self, current: f32) -> Result<(), Error> {
+        let dac_code = DacCode::try_from((current, self.channel))?;
+        self.write(dac_code.0 << 4, DAC_ADDR::DAC_DATA);
```

Comment:
Avoid the 4 bit left shift and just pretend the dac has 24 bits in `impl From<(channel, current)>`.

Reply:
How did you resolve this?

Reply:
Oh sry I accidentally clicked resolved..